packages/contractkit/src/identity/metadata.ts

```diff
@@ -0,0 +1,121 @@
+  TWITTER = 'TWITTER',
+}
+
+const UrlType = t.string
+const SignatureType = t.string
+const TimestampType = t.number
+
+const AttestationServiceURLClaimType = t.type({
+  type: t.literal(ClaimTypes.ATTESTATION_SERVICE_URL),
+  timestamp: TimestampType,
+  url: UrlType,
+})
+
+const DnsClaimType = t.type({
+  type: t.literal(ClaimTypes.DNS),
+  timestamp: TimestampType,
+  domain: t.string,
+})
+
+const NameClaimType = t.type({
+  type: t.literal(ClaimTypes.NAME),
+  timestamp: TimestampType,
+  name: t.string,
+})
+
+export const ClaimType = t.union([AttestationServiceURLClaimType, DnsClaimType, NameClaimType])
+export const SignedClaimType = t.type({
+  payload: ClaimType,
+  signature: SignatureType,
+})
+
+export const IdentityMetadataType = t.type({
+  claims: t.array(SignedClaimType),
+})
+
+export type SignedClaim = t.TypeOf<typeof SignedClaimType>
+export type AttestationServiceURLClaim = t.TypeOf<typeof AttestationServiceURLClaimType>
+export type DnsClaim = t.TypeOf<typeof DnsClaimType>
+export type NameClaim = t.TypeOf<typeof NameClaimType>
+export type IdentityMetadata = t.TypeOf<typeof IdentityMetadataType>
+export type Claim = AttestationServiceURLClaim | DnsClaim | NameClaim
+
+type ClaimPayload<K extends ClaimTypes> = K extends typeof ClaimTypes.DNS
+  ? DnsClaim
+  : K extends typeof ClaimTypes.NAME ? NameClaim : AttestationServiceURLClaim
+
+const isOfType = <K extends ClaimTypes>(type: K) => (
+  data: SignedClaim['payload']
+): data is ClaimPayload<K> => data.type === type
+
+export class IdentityMetadataWrapper {
+  data: IdentityMetadata
+
+  static async fetchFromURL(url: string) {
+    const resp = await fetch(url)
+    if (!resp.ok) {
+      throw new Error(`Request failed with status ${resp.status}`)
+    }
+    return this.fromRawString(await resp.text())
+  }
+
+  static fromRawString(rawData: string) {
+    const data = JSON.parse(rawData)
+    // TODO: We should validate:
+    // 1. data.claims being an array
+    // 2. payload being JSON-parsable
+    // This is hard to put into io-ts + we need to eventually do signature checking
+    return new IdentityMetadataWrapper({
+      claims: data.claims.map((claim: any) => ({
+        payload: JSON.parse(claim.payload),
+        signature: claim.signature,
+      })),
+    })
+  }
+
+  constructor(data: IdentityMetadata) {
+    const result = IdentityMetadataType.decode(data)
+    if (isLeft(result)) {
+      // TODO: We could probably return a more useful error in the future
+      throw new Error(PathReporter.report(result).join(', '))
+    }
+    this.data = result.right
+  }
+
+  get claims() {
+    return this.data.claims
+  }
+
+  toString() {
+    return JSON.stringify({
+      claims: this.data.claims.map((claim) => ({
+        payload: JSON.stringify(claim.payload),
+        signature: claim.signature,
+      })),
+    })
+  }
+
+  addClaim(claim: Claim) {
+    this.data = { claims: [...this.data.claims, this.signClaim(claim)] }
```

Comment:
This seems nor one thing, nor the other. We have a mutable object IdentityMetadataWrapper, but we treat `this.data` as an immutable one.

You can do:
```ts
this.data.claims.push(this.signClaim(claim))
```
and assume it's mutable and you "own" `this.data`
or treat everything immutable and then do
```ts
addClaim(claim: Claim): IdentityMetadataWrapper {
   return new IdentityMetadataWrapper({ claims: [...this.data.claims, this.signClaim(claim)] })
}
```
assuming of course you can create `IdentityMetadataWrapper` with it (which is another suggestion)

Reply:
Good point!

I'll do the mutation, but will probably leave `addClaim` as it is. I don't think chaining claims is going to be that common